### PR TITLE
Allow volume control via a StreamMagic streamer

### PR DESCRIPTION
**Pull request by ianparkinson**
Allows volume control via a StreamMagic streamer. Not enabled by default - the streamer's friendly name should be set with the `--amplifier` flag.

StreamMagic streamers - at least, the CXNv2 - support two different modes of controlling the volume. See https://www.cambridgeaudio.com/row/en/blog/getting-most-your-cxn-v2-pt2-volume-control. In pre-amp mode, the streamer acts as a digital pre-amp, with all the usual volume controls available - the volume is reported back to Vibin, and can be set to a specific value. Muting works, too. Otherwise, if the control bus is enabled, the streamer can control an attached Cambridge Audio amp such as the CXA series. In this case, volume can only be nudged up and down and no current volume level is reported back. If neither pre-amp nor control bus is enabled, no volume controls are available.

To allow the Vibin UI to handle both cases, this adds a list of supported actions to the AmplifierState object.

I'll also make a counterpart pull request for the client.

### CODE REVIEW

**Comment by ianparkinson** [comment]
I've added a client-side patch at https://github.com/mjoblin/vibinui/pull/261

You can test the UI with your CXNv2 from the settings page in the StreamMagic app - look for the the PreAmp and Control Bus settings. The UI will update as the configuration changes. You won't of course be able to test that the control bus mode actually works unless you have a Cambridge amp. I've tested this with my CXA61.

There are other ways of structuring this, and you may have other ideas :)

I've added a list of supported actions to AmplifierState - I needed *some* flag to indicate whether the amp supports bumping the volume up and down and, in the end, it seemed cleanest to add flags for all the possible actions. This is similar to the existing `active_controls` list in `TransportState`. I haven't yet updated other parts of the UI (e.g. `VolumeControl.tsx`, `SettingsMenu.tsx`) to use this actions list but I'm happy to follow up with those cleanups if you're happy with this approach.

When we're using the streamer as an amplifier, I'd like to drop the amplifier section in `SettingsMenu.tsx` and the popups that tell the user when the amplifier's power status has changed. The current approach doesn't give a *great* way of doing that (the easiest thing to do would be to check for the `power` action, and if it's unavailable, don't show those parts of UI). We could flow an extra flag somewhere (either `AmplifierState` or `SystemState`) saying that the amplifier and streamer are really the same device.

**Comment by mjoblin** [comment]
This is another fantastic add. Thank you! I tried it out (as much as I can without a Cambridge amp) and it works great. I put my CXNv2 into Pre-Amp mode and the volume controls popped up and behaved perfectly.

I noticed a couple of things. I'm not bothered by these at all, but want to mention them in case they're new to you:

* The volume arc shows the current volume (in blue) and the max available volume (in light grey). When in PreAmp mode on my CXNv2, the max volume (light grey) seemed to be locked to 50%, regardless of what the streamer's "Volume Limit" setting was set to. I'm not concerned about respecting that setting, although Vibin does currently respect it for Hegel amps. This might have just been a side-effect of me not actually using the streamer as a preamp. I defer to you on whether to deal with any of that.
* When I set the Control Bus to "Amplifier" (without an actual amp connected) the UI shows two volume icons next to each other (and no volume arc). This isn't a big deal, but wanted to mention it in case you'd like it to behave a more cleanly in this case.

One more thing before I forget: In the past, I learned that some Cambridge preamps seem to support 0-100 (percent) volume settings; while others use what seems to be some sort of volume stepping (with 30 steps total, from memory). Only mentioning that in case it triggers any thoughts in you. There did seem to be a difference in behavior between the CXNv2 and MXN10. (And I couldn't find a StreamMagic payload which would specify which flavor was preferred, although it's likely I just missed it or misunderstood something).

I haven't looked at the actual code yet, but I plan to do that soon. Without looking at the code though, overall I agree with your thinking about actions in AmplifierState, and dropping the amplifier settings when the amp and streamer are the same. I'll have more comments once I've looked more closely.

Thanks again for the work on this! It looks super cool. I'll update again soon.

**Comment by ianparkinson** [comment]
Hi - thanks for taking a look! Please do take your time reviewing, as I'm not going to be able to work on this for a couple of weeks anyway.

> The volume arc shows the current volume (in blue) and the max available volume (in light grey). When in PreAmp mode on my CXNv2, the max volume (light grey) seemed to be locked to 50%, regardless of what the streamer's "Volume Limit" setting was set to.

In VibinUI's Status page, there's a "Maximum amplifier volume" box which defaults to 50% - I think this is the locked value that you're seeing? I had to change that value to 100% for Vibin to use the whole range of volume values (apologies, I should have mentioned that). Is that value purely local to the UI? I couldn't see any way of syncing it up to the server or the amp.

The streamer's own max volume is reported via `smoip/zone/audio` so we should be able to use that. However, I've just tried it, and the volume limit on the CXNv2 doesn't actually seem to do anything! I can still set the volume above the streamer's limit via the app, the remote, or Vibin.

> When I set the Control Bus to "Amplifier" (without an actual amp connected) the UI shows two volume icons next to each other (and no volume arc). 

That's intentional on my part - but you may have better ideas :) In this case the streamer doesn't seem to know the actual volume level set on the amp. As far as I can tell, it can do is pass volume up/down commands up to the amp. The StreamMagic app (at least, on Android) only shows volume up/down controls in this case. For VibinUI, the volume arc, volume slider and mute button are all useless in this mode so I thought it'd be cleaner to just show buttons for the two available operations.

> One more thing before I forget: In the past, I learned that some Cambridge preamps seem to support 0-100 (percent) volume settings; while others use what seems to be some sort of volume stepping (with 30 steps total, from memory).

Ah, that's interesting. The `/zone/state` payload (for my CXNv2, when in pre-amp mode) looks like:

```
{
"zone": "ZONE1",
"data": {
"source": "MEDIA_PLAYER",
"power": true,
"pre_amp_mode": true,
"pre_amp_state": "on",
"mute": false,
"volume_step": 27,
"volume_percent": 90,
"volume_db": -3,
"cbus": "amplifier"
}
}
```

...where the `volume_step` is between 0 and 30. Changing the volume via `volume_step` or `volume_percent` both work and all three volume values are updated. Annoyingly, the display on the CXNv2 shows the `step` value, while the StreamMagic app shows the `percent` value.

So perhaps we can use either? Maybe some streamers only have `volume_step`? I chose to use the `percent` value because it matches the value used in Vibin UI, but perhaps the code should fall back to `volume_step` if `volume_percent` isn't reported.

Incidentally, I'm now seeing a bug in pre-amp mode - if I set the volume to some value that isn't a multiple of 10, after a couple of seconds, the volume reduces once or twice until it hits a multiple of 10. This only happens when VibinUI is running. I'll try to work out what's happening there.

Thanks again for taking the time to look at this.

**Comment by mjoblin** [comment]
Nice, thanks for the heads-up on your schedule. I plan to review the PR before you're able to get back to it, but please feel free to ping me here if you find yourself waiting for an update.

> In VibinUI's Status page, there's a "Maximum amplifier volume" box which defaults to 50% - I think this is the locked value that you're seeing?

Yep that's the one.

> Is that value purely local to the UI? I couldn't see any way of syncing it up to the server or the amp.

Up until now (i.e. for Hegel amps) it was purely local to the UI. But now...

> The streamer's own max volume is reported via smoip/zone/audio so we should be able to use that.

Sounds like a plan!

> However, I've just tried it, and the volume limit on the CXNv2 doesn't actually seem to do anything! I can still set the volume above the streamer's limit via the app, the remote, or Vibin.

Ha, OK maybe not a plan then :).

The max volume limit was put in there only to reduce the chance of blowing speakers out if the slider was accidentally dragged too far (it's much easier to over-drag a slider in a web browser than it is to bump a volume control on a remote). So this feature isn't strictly required. We have the following options (let me know if I'm missing any):

1. Ignore this feature for StreamMagic devices in preamp/cbus mode.
2. Try to respect the max volume specified in `smoip/zone/audio`.
3. Keep the max volume local to the UI and ignore `smoip/zone/audio`.

My vote would be for number 3, especially if even the Cambridge apps seem to ignore the `smoip/zone/audio` setting. In theory this shouldn't require any changes as the `<Slider>` in `VolumeControl.tsx` is already set to `max={amplifierMaxVolume}`.

No worries if you just fall back on number 1 though, and just ignore the max volume entirely. 

> As far as I can tell, it can do is pass volume up/down commands up to the amp. The StreamMagic app (at least, on Android) only shows volume up/down controls in this case.

That makes sense now! I thought it was a bug, but I think I was just confused by the chosen icons. From the [Tabler icon set](https://tabler.io/icons), I'm more partial to these (filled or unfilled): `square-rounded-arrow-up` or `square-rounded-plus` (and the `arrow-down`/`minus` equivalents). Please feel free to offer other suggestions if you're not a fan of those. If you're totally ambivalent then let's roll with the up/down arrows (filled, to match the play/pause/etc style).

As for volume:

> So perhaps we can use either? Maybe some streamers only have volume_step? I chose to use the percent value because it matches the value used in Vibin UI, but perhaps the code should fall back to volume_step if volume_percent isn't reported.

Here's what I've seen before: always set the volume with `step` and never with `percent`. It seems that some streamers support 30 steps while others support 100 steps. For the streamers that support 100 steps, the reported `step` and `percent` values will always be the same. For streamers that only support 30 steps, the reported `step` and `percent` values will always be different (except for the 0 case). This seems to work well.

In other words, the maximum possible volume step (ignoring UI volume limits) will be either 30 or 100. This could be used to set the max value of the volume slider. To determine whether to set the max to 30 or 100, you can check the latest reported `step` and `percent` values and if they're different then the max step value is 30 otherwise it's 100. (If both are 0; or no values have been received yet, then you could just fall back on 100).

I hope that makes sense. This _might_ help explain the pre-amp bug you noticed too.

Also, totally unrelated, I just ran the vibin back-end again to dig into the volume stuff and noticed another bug (completely unrelated to your PR). When starting vibin while the streamer is playing content without a playhead (e.g. internet radio), it freaks out about a missing `data` key. So thanks for indirectly making me aware of that! I'll make an issue so I don't forget to fix it.

**Comment by mjoblin** [comment]
I just had a play with some volume nudge icon options to see what felt nice. I'm curious what you think. These are just possibilities and I'm not suggesting that we have to choose one of these -- I'd love to see other suggestions if you have any.

volume-arrow-filled

volume-arrow

volume-plus-filled

volume-plus

Of these options I'm personally leaning towards the filled ones, and specifically the filled plus/minus -- but that's not a strong preference.

If you want to see these in the app then edit `VolumeUpDownControl.tsx`:

```typescript
import {
    IconSquareRoundedMinus,
    IconSquareRoundedMinusFilled,
    IconSquareRoundedPlus,
    IconSquareRoundedPlusFilled,
    IconSquareRoundedArrowDown,
    IconSquareRoundedArrowDownFilled,
    IconSquareRoundedArrowUp,
    IconSquareRoundedArrowUpFilled,
    IconVolume,
    IconVolume2,
} from "@tabler/icons-react";
```

Arrows (filled):

```typescript
        <Flex gap={1} align="center">
            <ActionIcon
                disabled={isAmpOff}
                sx={STYLE_DISABLEABLE}
                onClick={() => amplifierVolumeDown()}
            >
                <IconSquareRoundedArrowDownFilled
                    style={{ color: colorStandard }}
                />
            </ActionIcon>
            <ActionIcon
                disabled={isAmpOff}
                sx={STYLE_DISABLEABLE}
                onClick={() => amplifierVolumeUp()}
            >
                <IconSquareRoundedArrowUpFilled
                    style={{ color: colorStandard }}
                />
            </ActionIcon>
        </Flex>
```

Arrows:

```typescript
        <Flex gap={1} align="center">
            <ActionIcon
                disabled={isAmpOff}
                sx={STYLE_DISABLEABLE}
                onClick={() => amplifierVolumeDown()}
            >
                <IconSquareRoundedArrowDown
                    stroke={1.5}
                    color={colorStandard}
                />
            </ActionIcon>
            <ActionIcon
                disabled={isAmpOff}
                sx={STYLE_DISABLEABLE}
                onClick={() => amplifierVolumeUp()}
            >
                <IconSquareRoundedArrowUp
                    stroke={1.5}
                    color={colorStandard}
                />
            </ActionIcon>
        </Flex>
```

Plus/minus (filled):

```typescript
        <Flex gap={1} align="center">
            <ActionIcon
                disabled={isAmpOff}
                sx={STYLE_DISABLEABLE}
                onClick={() => amplifierVolumeDown()}
            >
                <IconSquareRoundedMinusFilled
                    style={{ color: colorStandard }}
                />
            </ActionIcon>
            <ActionIcon
                disabled={isAmpOff}
                sx={STYLE_DISABLEABLE}
                onClick={() => amplifierVolumeUp()}
            >
                <IconSquareRoundedPlusFilled
                    style={{ color: colorStandard }}
                />
            </ActionIcon>
        </Flex>
```

Plus/minus:

```typescript
        <Flex gap={1} align="center">
            <ActionIcon
                disabled={isAmpOff}
                sx={STYLE_DISABLEABLE}
                onClick={() => amplifierVolumeDown()}
            >
                <IconSquareRoundedMinus
                    stroke={1.5}
                    color={colorStandard}
                />
            </ActionIcon>
            <ActionIcon
                disabled={isAmpOff}
                sx={STYLE_DISABLEABLE}
                onClick={() => amplifierVolumeUp()}
            >
                <IconSquareRoundedPlus
                    stroke={1.5}
                    color={colorStandard}
                />
            </ActionIcon>
        </Flex>
```

Note:

* I don't know why the `Filled` icons only seem to respond to `style={{ color: colorStandard }}` and not `color={colorStandard}`, but hey, the former seems to work at least.
* I changed the flex gap to `1` as it made the buttons feel more closely related without feeling too cramped; but I'm definitely open to feedback on that.
* I removed the `size` prop as I preferred the default size for this use case.

File: vibin/models.py [hunk 1]
```diff
@@ -140,6 +141,7 @@ class MediaServerState(UPnPDeviceState):
 class AmplifierState(UPnPDeviceState):
     """Amplifier hardware state."""
 
+    actions: list[AmplifierAction]
```
[the remark this answers is not in the document]
Comment (ianparkinson) [reply vibin/models.py]: Renamed `supported_actions` - I agree this is subtly different from the "active" actions on the transport, which change with normal usage. "supported" keeps the sense that these actions are unlikely to change unless the user's configuration changes.

File: vibin/amplifiers/streammagic.py [hunk 2]
```diff
@@ -0,0 +1,269 @@
...
+            self._on_connect()
+
+    def _handle_state_message(self, message: Data):
+        try:
+            parsed = json.loads(message)
+        except (KeyError, json.decoder.JSONDecodeError):
+            return
+
+        if parsed["path"] != "/zone/state" or parsed["type"] != "update":
+            pass
```
[the remark this answers is not in the document]
Comment (ianparkinson) [reply vibin/amplifiers/streammagic.py]: Thanks, done.

**Comment by ianparkinson** [comment]
> 3. Keep the max volume local to the UI and ignore smoip/zone/audio.
My vote would be for number 3, especially if even the Cambridge apps seem to ignore the smoip/zone/audio setting. In theory this shouldn't require any changes as the <Slider> in VolumeControl.tsx is already set to max={amplifierMaxVolume}.

That's how the code currently behaves, isn't it? The max volume is kept local and honoured by the slider and volume arc. I *think* the default is 50% which causes a usability issue here, perhaps we can revisit that?

> Here's what I've seen before: always set the volume with step and never with percent.

Done. It now reads the maximum step value from `/zone/state/spec` so this should tolerate streamers with other max values. In practice, with a max step of 30, the slider is still a little clunky. How hard do you think it would be to flow down a `max_volume` value and have the slider show values between 0 and `max_volume` instead of between 0 and 100?

> This might help explain the pre-amp bug you noticed too.

Oh yes indeed, that was fun. See my comment at https://github.com/mjoblin/vibin/pull/125/commits/6fd4c3d7633d28daad095bdf71a819d3871c06cb.

> Also, totally unrelated, I just ran the vibin back-end again to dig into the volume stuff and noticed another bug (completely unrelated to your PR). When starting vibin while the streamer is playing content without a playhead (e.g. internet radio), it freaks out about a missing data key. 

I think that was on me - you had a catch-all for `KeyError` that suppressed this error, which I had inadvertently removed. I've restored it in this pull request.

> I just had a play with some volume nudge icon options

Thanks! I also prefer the filled -/+ icons, that should be reflected now in the other PR https://github.com/mjoblin/vibinui/pull/261

> I don't know why the Filled icons only seem to respond to style={{ color: colorStandard }} and not color={colorStandard}, but hey, the former seems to work at least.

It looks like, only for the filled icons, the SVG from tabler has `<path ... fill="currentColor">`. But I'm not sure if the CSS color can be set via properties in tsx - the `color` attribute actually sets `stroke`. Oh well, I'm neither a tsx nor a tabler expert, and I've kept your workaround of setting a `style` attribute.

Thanks again!

**Comment by mjoblin** [comment]
> That's how the code currently behaves, isn't it? The max volume is kept local and honoured by the slider and volume arc. I think the default is 50% which causes a usability issue here, perhaps we can revisit that?

Yup, correct. Your work in adding a new amplifier implementation has revealed a few usability issues in the current design, so I agree that revisiting some of this stuff makes a lot of sense.

I plan to merge these PRs as-is, but we could document some thoughts in an issue on how to improve volume handling. (And of course I'd be more than happy if you wanted to take it on; but no expectations there, and I'd still want to merge the current work).

> How hard do you think it would be to flow down a max_volume value and have the slider show values between 0 and max_volume instead of between 0 and 100?

That should be very doable. One clarification though: I believe the current slider's max is set to the soft max volume set in the UI's settings (the one that defaults to 50%), rather than the max attainable volume on the amp itself. Thinking out loud a little here:

* Yep I think the amplifier's max volume (usually 100, but maybe 30 or some other int value) should be available to the UI.
* Some amps may not announce their max volume, but worst case the `Amplifier` implementation could hard code it.
* We'd have to decide whether vibin should ever respect an amplifier's soft max setting (if the amplifier supports that feature).
* Vibin currently makes an attempt (mostly in the back-end I think) to normalize the volume by using floats (where 1 is the max, which could map to 100 or 30 or anything else). You suggested swapping this out for an int in the other PR and I agree. I still like the normalization benefit of using a float; but it introduces the problems you mentioned in the other PR. And using ints would at least ensure a perfect 1:1 between what Vibin displays for volume and what the amp displays on its screen. Using ints ends up exposing an amp implementation detail to the UI; but in this case it seems warranted.

Othertwise, the filled -/+ icons are a nice choice -- definitely agreed. And nice work tracking down what was happening in 6fd4c3d; there's some crazy sauce going on in there for sure!

**Pull request merged**